ray_tile_and_stitch_util.py

```diff
@@ -0,0 +1,341 @@
+        # don't want moving window to be larger than row size of input raster
+        if i + block_size < ysize:
+            rows = block_size
+        else:
+            rows = ysize - i
+
+        # read col
+        for id_j, j in enumerate(x_list):
+            if j + block_size < xsize:
+                cols = block_size
+            else:
+                cols = xsize - j
+            # get block out of the whole raster
+            # todo check the array values is similar as ReadAsArray()
+            band_1_array = final_array_4[i: i + rows, j: j + cols]
+            band_2_array = final_array_2[i: i + rows, j: j + cols]
+            band_3_array = final_array_3[i: i + rows, j: j + cols]
+
+            # filter out black image
+            if (
+                band_3_array[0, 0] == 0
+                and band_3_array[0, -1] == 0
+                and band_3_array[-1, 0] == 0
+                and band_3_array[-1, -1] == 0
+            ):
+                continue
+
+            # stack three bands into one array
+            img = np.stack((band_1_array, band_2_array, band_3_array), axis=2)
+            cv2.normalize(img, img, 0, 255, cv2.NORM_MINMAX)
+            img = img.astype(np.uint8)
+            B, G, R = cv2.split(img)
+            out_B = cv2.equalizeHist(B)
+            out_R = cv2.equalizeHist(R)
+            out_G = cv2.equalizeHist(G)
+            final_image = np.array(cv2.merge((out_B, out_G, out_R)))
+
+            # Upper left (x,y) for each images
+            ul_row_divided_img = uly + i * y_resolution
+            ul_col_divided_img = ulx + j * x_resolution
+
+            tile_metadata = ImageTileMetadata(
+                upper_left_row=ul_row_divided_img, upper_left_col=ul_col_divided_img, tile_num=tile_count, id_i=id_i, id_j=id_j)
+            image_tile = ImageTile(
+                tile_values=final_image, tile_metadata=tile_metadata)
+            tiles.append(image_tile)
+            tile_count += 1
+
+    # --------------- Store all the title as an object file
+    image_metadata = ImageMetadata(
+        len_x_list=len(x_list), len_y_list=len(y_list), x_resolution=x_resolution, y_resolution=y_resolution)
+    row["image_metadata"] = image_metadata
+    new_rows = []
+    tile_count = 0
+    for image_tile in tiles:
+        new_row = copy.deepcopy(row)
+        new_row["image_tile"] = image_tile
+        new_row["tile_num"] = tile_count
+        new_rows.append(new_row)
+        tile_count += 1
+    return new_rows
+
+
+def stitch_shapefile(group: pd.DataFrame):
+    """
+    Create a shapefile for each image.
+    Note that normally ray rows are dictionaries but this is a group because we've called
+    grouped by. When ray does a groupby and the rows can't be represented in numpy arrays
+    it uses pandas dataframe.
+
+    Parameters
+    ----------
+    group: a pandas dataframe that has all of the shapefile information for each tile in the
+            image.
+    """
```

Comment:
This documentation is extremely helpful! It would be even better if we could add some information around the output as well since it's not entirely clear why we return the "first_row" of the group.

Reply:
ack done :)